routing/graph.go

```diff
@@ -1,4 +1,201 @@
 package routing
 
+import (
+	"fmt"
+
+	"github.com/boltdb/bolt"
+	"github.com/lightningnetwork/lnd/channeldb"
+	"github.com/roasbeef/btcd/btcec"
+)
+
+const (
+	// TotalConcurrentThreads is the maximum number of processes
+	// to run in parallel for asynchronous functions.
+	TotalConcurrentThreads = 20
+)
+
 // TODO(roasbeef): abstract out graph to interface
 //  * add in-memory version of graph for tests
+
+// TODO(nalinbhardwaj): move routing.ChannelGraphSource here.
+
+// Vertex is a simple alias for the serialization of a compressed Bitcoin
+// public key.
+type Vertex [33]byte
+
+// NewVertex returns a new Vertex given a public key.
+func NewVertex(pub *btcec.PublicKey) Vertex {
+	var v Vertex
+	copy(v[:], pub.SerializeCompressed())
+	return v
+}
+
+// String returns a human readable version of the Vertex which is the
+// hex-encoding of the serialized compressed public key.
+func (v Vertex) String() string {
+	return fmt.Sprintf("%x", v[:])
+}
+
+// Edge is a struct holding the edge info we need when performing path
+// finding using a graphSource.
+type Edge struct {
+	edgeInfo *channeldb.ChannelEdgeInfo
+	outEdge  *channeldb.ChannelEdgePolicy
+	inEdge   *channeldb.ChannelEdgePolicy
+}
+
+// graphSource is a wrapper around a graph that can either be in memory or
+// on disk, like the regular channeldb.ChannelGraph.
+type graphSource interface {
+	ForEachNode(func(Vertex) error) error
+	ForEachChannel(Vertex, func(*Edge) error) error
+
+	// Async versions of same functions
+	AsyncForEachNode(func(Vertex, chan error)) error
```

Comment:
Since it is not actually getting each node/vertex that is slow, the concurrency handling shouldn't be handled at this level of the abstraction. Instead the graph source should just be having the sync methods, and the caller be responsible for launching the necessary goroutines in his preferred mode. E.g. like this:

```golang
vertices := make(chan Vertex)
for i := 0; i < numGoRoutines; i++ {
       go func() {
          for {
              vertex, ok := <- vertices
              if !ok {
                   return
              }
              doSlowCalculation(vertex)
            }
       }()
}
g.ForEachNode(func(v Vertex) error {
        vertices <- v
})
close(vertices)
```

Reply:
Neat! Didn't think of this. Fixing.